### ISSUE TITLE
NaN in the prediction array

### ISSUE THREAD

**Issue by MRLoghmani** [fchollet/keras#2967]
Does anyone have an idea of how a NaN can rise in the prediction array, that is in the return value of predict() method? I check the data with which I'm feeding my net, but everything seems to be fine. 
If it can be helpful I'm training a recurrent net with some (2 to 4) GRU layers and a TimeDistributedDense layer at the end to perform a multi-class classification task.

**Comment by hidayata** [comment]
Any luck with this? I'm getting NaNs in my prediction array with a 4-layer dense network.

**Comment by budhiraja** [comment]
Having the same issue. Any luck?

**Comment by minesh1291** [comment]
Why this issue is closed..!!

**Comment by Aaron-Edwards-93** [comment]
I don't suppose anybody ever found a solution to this? I'm having this problem using a recurrent network (with GRU's). I am only trying to predict one record at a time, and the data looks fine to me, but no matter what I do the predict function returns nan!

**Comment by budhiraja** [comment]
There was a problem with my data : it had NaN values.

**Comment by zyavrik** [comment]
> There was a problem with my data : it had NaN values.

It's not the case for me. Predictions just show NaN values without any reason...

**Comment by bojanpotocnik** [comment]
I also have this problem. Anyone found a solution? I checked the input data and it is fine, all is isfinite() - Ridge and Lasso Regressions work fine with the same data. However Sequential model finishes fit() without any warnings but then outputs nan for any predict operation.

**Comment by bojanpotocnik** [comment]
I had the same problem today. Try to use different optimizer (Adam or SGD with momentum instead of SGD), for me replacing SGD with Adam solved the problem.

**Comment by dickreuter** [comment]
Same problem here. The data is not the problem as I cannot reproduce it on a different machine. Could it be a memory issue with tensorflow?

**Comment by awhillas** [comment]
i'm getting this but only after I load the model from file. I predict before i save and it works fine :-?

**Comment by jhmlam** [comment]
In the same situation here. It is very strange, just as @awhillas said, training and prediction was fine before saving. Training is fine after saving too. But strangely, prediction turns Nan even though using the same training set....

**Comment by cogons** [comment]
@awhillas @jhmlam stuck in the same situation here. Things went well before I tried to save the model. After that, the training was fine, but the prediction returns list of NAN. I cant reproduce it on the other machine.

**Comment by tadaoisgod** [comment]
I'm in the same situation. Fine-tuning a ResNet model, runs fine while training, predicts correctly after training stops, but after load/save everything turns to NaN.

**Comment by cgodfreyUW** [comment]
Having exactly the same problem - after save/load, get lots of NaN predictions. 

Also wondering why this issue is closed ...

**Comment by jaungiers** [comment]
Just stumbled across this problem also when save/load'ing a model, predictions are NaNs. 
Model is a 3-layer recurrent network: 2 GRU layers and a TimeDistributed Dense layer with an Adam optimizer.

**Comment by Jahanzaib-int** [comment]
After very hard efforts by going through all the steps and verifying them all, I discovered that there is Infinity in my data. My dataset was in .csv and very large in amount, so I had to write python script to remove all such values. **Yes 1 more thing you should consider removing these values like 2 or 3 times as the loop breaks immaturely after row count is less but in the buffer there might still be some missed values that are not parsed.**

//Import your dataset
dataset = np.array(dataset)
count = 0
rows = len(dataset)
columns = len(dataset[1])
for x in range(0,rows):
    for y in range(0,columns):
        if(dataset[x][y] in {'NaN','infinity','nan','Infinity'}):
            dataset = np.delete(dataset,x,0)
            count = count +1
print("Deleted Rows: "+str(count))

**Comment by xiormeesh** [comment]
In most cases it's caused by exploding gradients: make sure you normalize numerical values or clip gradients

**Comment by vedantamvamsi** [comment]
reloading the model worked for me

**Comment by MedhaviMonish** [comment]
same is happening to me , i am training for reinforcement learning in continuous action space , 
so as Jahanzaib-int suggested i checked the observation then yes the observation has nan in it which was causing the problem

**Comment by soroushhashemifar** [comment]
I had the same problem in deep q learning algorithm. After spending a couple of hours debugging the code, I found that something was making reward, inf, so as the labels the model was training on. I resolved the inf problem and now the model works well. PAY TOO MUCH ATTENTION TO INPUTS AND OUTPUTS OF THE MODEL WHILE FITTING, it's more likely to find something wrong there.

**Comment by Bahador-Bakhshi** [comment]
I had the same problem. Changing the optimizer from SGD to the RMSprop resolved the issue.

**Comment by kvantricht** [comment]
I had the same problem. All fine during training, all NaN's when predicting after loading the model. Even a zeros-array returns NaN, so it's not about the inputs. Turns out I trained the model on GPU and did inference on CPU. GPU training "seemed" to go fine, although actually my RNN layers quickly got NaN weights. GPU doesn't care and moves on, eventually turning my network into a Dense network where only the layers after RNN had valid weights.
CPU inference gives NaN predictions when seeing the RNN NaN weights. So one thing to check: does your model contain NaN weights?
